Add three microphysics number concentrations to RRFS output

jaymes-kenyon opened the pull request:
Updates to allow RRFS output of cloud-water number concentration, cloud-ice number concentration, and rain number concentration on native coordinate surfaces, toward resolving issue #731

jaymes-kenyon committed:
(1) Updates to read-in the number concentrations of cloud ice, cloud water, and rain
(2) Slight reorganization of the "use vrbls3d" code section to re-group several microphysics-related variables

jaymes-kenyon committed:
Enabling RRFS output of cloud-ice number concentration (NCCICE), cloud-water number concentration (NCONCD), and rain number concentration (SPNCR) on native coordinate surfaces.

WenMeng-NOAA commented:
@jaymes-kenyon Do you have sample RRFS model output files in netcdf which include these three number concentration fields for testing?

EricJames-NOAA reviewed (approved):
Looks great!  Nice work, Jaymes!

jaymes-kenyon commented:
Hi @WenMeng-NOAA The RRFS_B (CONUS) has the required netCDF output in the "dyn" files.  I'm able to access these files on Jet, e.g., `/lfs4/BMC/nrtrr/NCO_dirs/v0.5.2/stmp/2023080713/fcst_fv3lam/ ` Is this helpful?

WenMeng-NOAA commented:
> Hi @WenMeng-NOAA The RRFS_B (CONUS) has the required netCDF output in the "dyn" files. I'm able to access these files on Jet, e.g., `/lfs4/BMC/nrtrr/NCO_dirs/v0.5.2/stmp/2023080713/fcst_fv3lam/ ` Is this helpful?

@jaymes-kenyon Can you transfer data files to Hera? The files (dyn*, phy*) at one forecast hour would be enough for testing.

jaymes-kenyon commented:
@WenMeng-NOAA — I placed some sample files on Hera at `/scratch1/BMC/wrfruc/kenyon/upp_development`.  This is the 1-h forecast, initialized at 2023080811.  Thanks!

WenMeng-NOAA commented:
@jaymes-kenyon You may adapt the definitions of ID 754 and 752 from  post_avblflds_raphrrr.xml to post_avblflds.xml as:
```
752:
change NCIP into NCCICE
754:
change NCRAIN into SPNCR
```

WenMeng-NOAA commented on parm/fv3lam_rrfs.xml:
```diff
@@ -2528,6 +2546,16 @@
+       <param>
+       <shortname>GSD_NCRAIN_ON_HYBRID_LVL</shortname>
+       <pname>SPNCR</pname> 
+       <table_info>NCEP</table_info>
```
@jaymes-kenyon This line should be deleted for properly generating SPNCR.

jaymes-kenyon replied:
Deleted the "table_info" line

WenMeng-NOAA commented on sorc/ncep_post.fd/INITPOST_NETCDF.f:
```diff
@@ -891,6 +894,12 @@ SUBROUTINE INITPOST_NETCDF(ncid2d,ncid3d)
        spval,recname(17),coarsepm(ista_2l,jsta_2l,1,1),lm)
        call read_netcdf_3d_para(ncid2d,im,jm,ista,ista_2l,iend,iend_2u,jsta,jsta_2l,jend,jend_2u, &
```
@jaymes-kenyon In the line 895, 897, 899, 901, change "ncid2d" into "ncid3d".

jaymes-kenyon replied:
Thanks!  The latest commit makes these changes.  I am just confirming that I modified line 895, which existed prior to this PR. This line affects the handling of the "ext550" netCDF array. @EricJames-NOAA

WenMeng-NOAA replied:
@jaymes-kenyon I checked RRFS history files. The variable "ext550" is from  file *phy*. So the line 895 should be kept as "nid2dd". Sorry for misleading. @EricJames-NOAA may chime in.

EricJames-NOAA replied:
@WenMeng-NOAA Yes I think line 895 should stay as ncid2d.  I think just lines 897, 899, and 901 should be ncid3d.  @jaymes-kenyon did you have a chance to test the new code for RRFS to see if the new fields come out as expected?

WenMeng-NOAA replied:
@EricJames-NOAA Thanks for confirming. I am working on a test with model files provided by @jaymes-kenyon and will have both of you to verify results soon.

jaymes-kenyon replied:
@WenMeng-NOAA @EricJames-NOAA — From my tests on Jet (CONUS output), the number-concentration fields look reasonable at k=20 (just a sample). But, we may wish to explore the GRIB2 precision settings later. I will revert line 895 ("ext550") to its previous form.  Thanks!

EricJames-NOAA replied:
The mean/max/min values of the new fields look reasonable with wgrib2.  Thanks for the test @WenMeng-NOAA!

WenMeng-NOAA replied:
@jaymes-kenyon Would you like the GRIB2 precision tuning is a separate PR or not? If yes, we could proceed the RT and  merging process with this PR.

jaymes-kenyon replied:
@WenMeng-NOAA — I would suggest updates to precision settings in a possible separate / future PR.  This will give NCAR collaborators a chance to look at this new data and see if it meets their requirements.  Feel free to merge this PR when ready, then we can install the updates into the real-time RRFS runs.

jaymes-kenyon committed:
Updates to static control files:
(1) In post_avblflds.xml, changing "NCIP" to "NCCICE" (ID 752)	and "NCRAIN" to "SPNCR" (ID 754).  This is consistent with the ID numbers used in post_avblflds_raphrrr.xml.
(2) Removing the "table_info" specification for "SPNCR" in fv3lam_rrfs.xml.
(3) Re-generated flat flie.

jaymes-kenyon committed:
In INITPOST_NETCDF.f, changing the "ncid2d" argument to "ncid3d" withinin several calls to "read_netcdf_3d_para".

WenMeng-NOAA commented:
@jaymes-kenyon and @EricJames-NOAA My UPP standalone test with model files provided by Jaymes is at
/scratch1/NCEPDEV/stmp2/Wen.Meng/post_rrfs_2023080811. Please let me know if you see issues.

jaymes-kenyon committed:
Reverting the read-in of "ext550" to its previous form.

AndrewBenjamin-NOAA commented:
I took a look at the stats in wgrib2 as well as a quick panoply plot and these look reasonable to me. @WenMeng-NOAA would you like me to go ahead and submit my review?

WenMeng-NOAA commented:
@EdwardColon-NOAA Please be aware of the update in RRFS control files in this PR. That will impact on 3DRTMA since RRFS and 3DRTMA share with the same control files.

WenMeng-NOAA commented:
The UPP RT were completed on WCOSS2, Hera and Orion. There will be baseline recreation for RRFS and 3DRTMA with this PR.

WenMeng-NOAA commented:
This PR is ready for merging.
